lib/src/layer/tile_layer/tile_image_manager.dart

```diff
@@ -0,0 +1,263 @@
+    }
+  }
+
+  void reloadImages(
+    TileLayer layer,
+    TileBounds tileBounds,
+  ) {
+    // If a TileImage's imageInfo is already available when load() is called it
+    // will call its onLoadComplete callback synchronously which can trigger
+    // pruning. Since pruning may cause removals from _tiles we must not
+    // iterate _tiles directly otherwise a concurrent modification error may
+    // occur. To avoid this we create a copy of the collection of tiles to
+    // reload and iterate over that instead.
+    final tilesToReload = List<TileImage>.from(_tiles.values);
+
+    for (final tile in tilesToReload) {
+      tile.imageProvider = layer.tileProvider.getImage(
+        tileBounds.atZoom(tile.coordinates.z).wrap(tile.coordinates),
+        layer,
+      );
+      tile.load();
+    }
+  }
+
+  void markAsNoLongerCurrentOutside(
+      int currentTileZoom, DiscreteTileRange noPruneRange) {
+    for (final entry in _tiles.entries) {
+      final tile = entry.value;
+      final c = tile.coordinates;
+
+      if (tile.current &&
+          (c.z != currentTileZoom ||
+              !noPruneRange.contains(CustomPoint(c.x, c.y)))) {
+        tile.current = false;
+      }
+    }
+  }
+
+  // Evicts error tiles depending on the [evictStrategy].
+  void evictErrorTiles(
+    DiscreteTileRange tileRange,
+    EvictErrorTileStrategy evictStrategy,
+  ) {
+    if (evictStrategy == EvictErrorTileStrategy.notVisibleRespectMargin) {
+      final toRemove = <String>[];
+      for (final entry in _tiles.entries) {
+        final tile = entry.value;
+
+        if (tile.loadError && !tile.current) {
+          toRemove.add(entry.key);
+        }
+      }
+
+      for (final key in toRemove) {
+        _remove(key, evictImageFromCache: (_) => true);
+      }
+    } else if (evictStrategy == EvictErrorTileStrategy.notVisible) {
+      final toRemove = <String>[];
+      for (final entry in _tiles.entries) {
+        final tile = entry.value;
+        final c = tile.coordinates;
+
+        if (tile.loadError &&
+            (!tile.current || !tileRange.contains(CustomPoint(c.x, c.y)))) {
+          toRemove.add(entry.key);
+        }
+      }
+
+      for (final key in toRemove) {
+        _remove(key, evictImageFromCache: (_) => true);
+      }
+    }
+  }
+
+  void prune(EvictErrorTileStrategy evictStrategy) {
+    for (final tile in _tiles.values) {
+      tile.retain = tile.current;
+    }
+
+    for (final tile in _tiles.values) {
+      if (tile.current && !tile.active) {
+        final coords = tile.coordinates;
+        if (!_retainAncestor(coords.x, coords.y, coords.z, coords.z - 5)) {
+          _retainChildren(coords.x, coords.y, coords.z, coords.z + 2);
+        }
+      }
+    }
+
+    final toRemove = <String>[];
+    for (final entry in _tiles.entries) {
+      if (!entry.value.retain) toRemove.add(entry.key);
+    }
+
+    for (final key in toRemove) {
+      _removeWithDefaultEviction(key, evictStrategy);
+    }
+  }
```

Comment:
Hi @rorystephenson, I currently working on some code improvements and wanted to ask if there is a reason to have the logic splitted into three loops here?

Reply:
Hi @josxha, it's great that you're digging in looking for a change to improve the code.

It looks like this code originates from [this commit](https://github.com/fleaflet/flutter_map/commit/d4aefaecc44d0c0b431375fd7bc97b2ef1a214e7).

I don't think we can reduce the for loops in the current form (more on that at the end). Tiles have an existing `retain` value (from a previous prune run) which might be set to true or false. Imagine if a tile, tile X, previously had a `retain` value of true but it is no longer current. In the current implementation we first set tile X's `retain` to the value of `current`, false, in the first for loop. Then we iterate through the tiles and, if we need to retain tile X because it is the ancestor/child of a current tile that is not loaded yet we will set its retain to `true`. 

If we combine those two for loops we might first encounter a non-loaded child of tile X when iterating and set its `retain` value to true before iterating over tile X and setting its `retain` back to false because it is not current.

We cannot combine the 2nd/3rd for loops because we need to iterate all tiles before we know that we do not need to retain a child (as mentioned before, it may be the child/parent of a non-loaded current tile).

Finally we can't combine the 3rd and 4th for loops because this would cause a concurrent modification exception. Calling _removeWithDefaultEviction removes an element from _tiles so we must first build a new list of removals from _tiles before iterating over the new list and performing the removals.

---

That said the fact that we have four loops to perform removals is a code smell and you're right to point it out. I have been experimenting with refactoring tile pruning to make the logic easier to follow and you have made me realise that I can probably remove an iteration in my [new implementation](https://github.com/rorystephenson/flutter_map/blob/306891d6194ed9a9a3625565724f38fa700c9eb1/lib/src/layer/tile_layer/tile_removal_state.dart#L52-L68). I need to finish my work on that before opening a PR though.

Reply:
Just r.e the removals. I never really liked the whole prune methodology, it felt kind of wrong, and I did have a dabble at a change, but it wasn't quite as smooth.

The basic premise was to not think about pruning tiles, but starting from scratch, "what tiles do we need to display immediately", if a tile is outstanding, see if there's a parent (or child) tile zoom that covers it (similar in concept to the retainParent/Child iirc), and use that.

It's a subtle difference mentally, if you wanted a peak [
](https://github.com/ibrierley/flutter_map_vector_tiles/blob/master/lib/VectorTileWidget.dart) (basically line 244-384). As mentioned, it didn't work as well so I wouldn't just copy it, but it may just give a slightly different angle on it..I wonder if my problem was that it wouldn't necessarily try and use a tile from the last frame, it would pick any as a backup). Don't want to distract the work you're doing tho, so don't get bogged down on that this. Maybe it should be a separate change.

Reply:
Thank you very much for this great explaination @rorystephenson!